src/stub/core_impl/tracepoints.rs

```diff
@@ -0,0 +1,304 @@
+                    // We may or may not have another action after our mask
+                    let mask = if let Some(mask_end) = mask_end {
+                        let (mask_bytes, next) = mask.split_at_mut(mask_end.0);
+                        unparsed = Some(next);
+                        decode_hex_buf(mask_bytes).ok()?
+                    } else {
+                        unparsed = None;
+                        decode_hex_buf(mask).ok()?
+                    };
+                    (f)(&TracepointAction::Registers {
+                        mask: ManagedSlice::Borrowed(mask),
+                    });
+                }
+                Some([b'M', _mem_args @ ..]) => {
+                    // Unimplemented: even simple actions like `collect *(int*)0x0`
+                    // are actually assembled as `X` bytecode actions
+                    return None;
+                }
+                Some([b'X', eval_args @ ..]) => {
+                    let mut len_end = eval_args.splitn_mut(2, |b| *b == b',');
+                    let (len_bytes, rem) = (len_end.next()?, len_end.next()?);
+                    let len: usize = decode_hex(len_bytes).ok()?;
+                    let (expr_bytes, next_bytes) = rem.split_at_mut(len * 2);
+                    let expr = decode_hex_buf(expr_bytes).ok()?;
+                    (f)(&TracepointAction::Expression {
+                        expr: ManagedSlice::Borrowed(expr),
+                    });
+                    unparsed = Some(next_bytes);
+                }
+                Some([b'-']) => {
+                    more = true;
+                    break;
+                }
+                Some([]) | None => {
+                    break;
+                }
+                _ => return None,
+            }
+        }
+
+        Some(more)
+    }
+}
+
+impl<T: Target, C: Connection> GdbStubImpl<T, C> {
+    pub(crate) fn handle_tracepoints(
+        &mut self,
+        res: &mut ResponseWriter<'_, C>,
+        target: &mut T,
+        command: Tracepoints<'_>,
+    ) -> Result<HandlerStatus, Error<T::Error, C::Error>> {
+        let ops = match target.support_tracepoints() {
+            Some(ops) => ops,
+            None => return Ok(HandlerStatus::Handled),
+        };
+
+        crate::__dead_code_marker!("tracepoints", "impl");
+
+        match command {
+            Tracepoints::QTinit(_) => {
+                ops.tracepoints_init().handle_error()?;
+                // GDB documentation doesn't say it, but this requires "OK" in order
+                // to signify we support tracepoints.
+                return Ok(HandlerStatus::NeedsOk);
+            }
+            Tracepoints::qTStatus(_) => {
+                let status = ops.trace_experiment_status().handle_error()?;
+                res.write_str("T")?;
+                res.write_dec(if status.running { 1 } else { 0 })?;
+                for explanation in status.explanations.iter() {
```

Comment:
We need to write the `T0`/`T1` running status before we write any of the explanations. If we pass the `res` writing callback to the target implementation, then it would need to have some mechanism of reporting the experiment state before it can run the explanation callback. We can't do like an `&mut FnOnce(ExperimentStatus)->TargetResult<&mut FnMut(ExperimentExplantion)->TargetResult<(),Self>, Self>` because we can't return a borrow from the closure due to lifetime issues, and can't return a trait object by value due to it being unsized.

Reply:
Thinking about it, one solution would be to split the API into `trace_experiment_status` and then `trace_experiment_statistics` or something? Does gdbstub particularly care about matching API surface 1:1 with gdb packets or would that be ok?

Reply:
> Does gdbstub particularly care about matching API surface 1:1 with gdb packets or would that be ok?

Reflecting the underlying packet structure to end users is actually an _anti-goal_ of `gdbstub` 😄

It just-so-happens that the many times, the protocol is relatively trivial, so the resulting API ends up matching the packets 1:1... but there are other cases where `gdbstub` goes out of its way to expose a more "user-friendly" API, and then take on whatever heavy-lifting it needs to do in order to map those friendly end-user semantics onto the underlying protocol.